### PR TITLE
[Speedy] recompute contract metadata when using different pkg versions 

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**

### CODE REVIEW

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala [hunk 1]
```diff
@@ -98,11 +94,12 @@ class UpgradeTest(majorLanguageVersion: LanguageMajorVersion)
     p""" metadata ( '-upgrade-test-' : '3.0.0' )
       module M {
 
-      record @serializable T = { sig: Party, aNumber: Int64, optSig: Option Party };
+      record @serializable T = { sig: Party, obs: Party, aNumber: Int64, optSig: Option Party };
       template (this: T) = {
         precondition True;
         signatories '-pkg1-':M:sigs (M:T {sig} this) (M:T {optSig} this);
-        observers Nil @Party;
+         observers '-pkg1-':M:sigs (M:T {obs} this) (None @Party);
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:101]:
```suggestion
        observers '-pkg1-':M:sigs (M:T {obs} this) (None @Party);
```

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala [hunk 2]
```diff
@@ -113,15 +110,37 @@ class UpgradeTest(majorLanguageVersion: LanguageMajorVersion)
     """
   }
 
+  val pkgId4: Ref.PackageId = Ref.PackageId.assertFromString("-pkg4-")
+  private lazy val pkg4 = {
+    // add an optional additional signatory
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:115]: This is already done by pkg3. I've stared some time at the screen but I can't tell the difference between pkg3 and pkg4. Are they the same?

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala [hunk 3]
```diff
@@ -322,20 +351,68 @@ class UpgradeTest(majorLanguageVersion: LanguageMajorVersion)
     }
   }
 
+  "upgrade" - {
+    "be able to fetch a same contract using different versions" in {
+      // The following code is not properly type, but emulate two commands that fetch a same contract using different versions.
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:356]:
```suggestion
      // The following code is not properly typed, but emulates two commands that fetch a same contract using different versions.
```

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala [hunk 4]
```diff
@@ -322,20 +351,68 @@ class UpgradeTest(majorLanguageVersion: LanguageMajorVersion)
...
+                 x1: Unit <- '-pkg2-':M:do_fetch cid;
+                 x2: Unit <- '-pkg3-':M:do_fetch cid
+               in upure @Unit ()
+          """,
+        ContractInstance(pkgName, i"'-pkg1-':M:T", v1_base),
+      )
+      res shouldBe a[Right[_, _]]
+    }
+    "do recompute and check immutability of meta data when using different versions" in {
+      // The following code is not properly type, but emulate two commands that fetch a same contract using different versions.
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:369]:
```suggestion
      // The following code is not properly typed, but emulates two commands that fetch a same contract using different versions.
```

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala [hunk 5]
```diff
@@ -40,19 +34,20 @@ class UpgradeTest(majorLanguageVersion: LanguageMajorVersion)
...
     p""" metadata ( '-upgrade-test-' : '1.0.0' )
     module M {
 
-      record @serializable T = { sig: Party, aNumber: Int64 };
+      record @serializable T = { sig: Party, obs: Party, aNumber: Int64 };
       template (this: T) = {
         precondition True;
         signatories M:sigs (M:T {sig} this) (None @Party);
-        observers Nil @Party;
+        observers '-pkg1-':M:sigs (M:T {obs} this) (None @Party);
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:49]: I was a bit confused as to why `observers` was using the `sigs` function at first. Maybe you could rename it to `mkList` or something like that for clarity?

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala (hunk 2 above)
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:115]:
```suggestion
    // swap signatories and observers
```

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala [hunk 6]
```diff
@@ -113,15 +110,37 @@ class UpgradeTest(majorLanguageVersion: LanguageMajorVersion)
...
   }
 
+  val pkgId4: Ref.PackageId = Ref.PackageId.assertFromString("-pkg4-")
+  private lazy val pkg4 = {
+    // add an optional additional signatory
+    implicit def pkgId: Ref.PackageId = pkgId4
+    p""" metadata ( '-upgrade-test-' : '4.0.0' )
+      module M {
+
+      record @serializable T = { sig: Party, obs: Party, aNumber: Int64, optSig: Option Party };
```
Comment (paulbrauner-da) [inline sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:120]: Is `optSig` unused on purpose?

**Review by paulbrauner-da — APPROVED** [review]
Thank you!

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala [hunk 7]
```diff
@@ -40,19 +34,20 @@ class UpgradeTest(majorLanguageVersion: LanguageMajorVersion)
   private[this] implicit def parserParameters(implicit
       pkgId: Ref.PackageId
   ): ParserParameters[this.type] =
-    ParserParameters(pkgId, languageVersion = majorLanguageVersion.dev)
+    ParserParameters(pkgId, languageVersion = LanguageMajorVersion.V2.dev)
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:37]: ok

File: sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala (hunk 5 above)
Comment (remyhaemmerle-da) [reply sdk/daml-lf/interpreter/src/test/scala/com/daml/lf/speedy/UpgradeTest.scala:49]: ok

**Pull request merged**